README
jQuery AnySlider 1.2
Copyright 2011 Jonathan Wilsson
http://jonathanwilsson.com/projects/jquery-anyslider/

Orginal code by Jacob Gube
http://sixrevisions.com/tutorials/javascript_tutorial/create-a-slick-and-accessible-slideshow-using-jquery/
 

Requires at least jQuery 1.5.

AnySlider is a jQuery plugin which allows you to create a slider for any kind of content.

If you have any questions or thoughts feel free to contact me.


Changelog

1.2
Finally added a speed option
Various code improvements

1.1.2
Enhanced the options to always show/hide arrows or just on hover
Minor code cleanup

1.1
New arrows
Added some options:
	Always show/hide arrows
	Enable/disable keyboard navigation
	Custom labels for the controls
Fixed a minor bug were keyboard codes had to be converted into a string.
Separated options from internal variables

1.0.1
Rewrote bits of the code to perform better

1.0
Initial release

CHANGELOG.md
## Changelog
### 2.1.0 (2015-11-21)
* Improved CommonJS support so tools like Browserify and webpack work.
* jQuery Plugin manifest file has been removed.
* The current slide now has the class "as-current-slide".
* Fixed an issue with vertical scrolling on mobile and tablets ([#22](https://github.com/jwilsson/AnySlider/issues/22)).
* All external resources in demos etc. are now included over https.
* Fixed some CSS issues in the demo code.

### 2.0.2 (2014-12-18)
* Another version bump for npm after bower.json update.

### 2.0.1 (2014-12-15)
* Version bump for npm after bower.json update.

### 2.0.0 (2014-08-14)
* AnySlider now features an API. See the docs for more details.
* Added support for Zepto.js.
* Increased minimum required jQuery version to 1.7.0 (was 1.5.0).
* True crossfade for slides.
* Added support for AMD.
* Changed the $.fn namespace name to `anyslider` (note casing).
* Changed `touch-action: none;` to `touch-action: pan-y;` in the default CSS.
* Removed the following options:
    * `delay`
    * `responsive` (AnySlider is now responsive by default)
* Renamed the following options:
    * `bullets` to `showBullets`
    * `rtl` to `reverse`
    * `keyboardNav` to `keyboard`

### 1.7.1 (2013-03-08)
* Fixed a bug where the keyboard navigation wouldn't work.

### 1.7.0 (2013-12-17)
* It's now possible to scroll while touching the slider.
* Added namespaces to all events.
* Removed the data-num attribute on navigation bullets.

### 1.6.2 (2013-09-09)
* Fixed various issues regarding the cloning of slides.

### 1.6.1 (2013-06-18)
* Slides will always be cloned again, only cloning when there's more than one slide caused too much trouble at this point.

### 1.6.0 (2013-06-03)
* Added touch support for Windows 8.
* In all callbacks, `this` will now refer to the slider-element instead of the AnySlider-object.
* Fixed an issue where the autoplay timer wasn't reset. Props [@VitaliiBlagodir](https://github.com/VitaliiBlagodir).
* Fixed an issue where a single slide wasn't shown.

### 1.5.2 (2012-04-27)
* Added a "delay" option to control the delay between fade slides.
* Fixed a bug where the navigation bullets weren't correctly updated when autoplaying.
* If there is only one slide in the slider, that slide won't be cloned.

### 1.5.1 (2012-18-12)
* Fixed a compatibility issue with jQuery 1.9.

### 1.5.0 (2012-11-18)
* AnySlider is now responsive!
* The option "showOnHover" has been removed. Arrows are only drawn if "showControls" is true. The ability to show/hide controls is now handled by CSS.
* The navigation bullets have been moved outside the slider element to allow for better control of the positioning.
* Other bugfixes, optimizations and improvements.

### 1.4.3 (2012-08-22)
* Fixed an issue where the slider would crash in older versions of jQuery.
* Fixed an issue where the navigation bullets weren't correctly updated.
* Added the following callbacks:
    * beforeChange which triggers before a slide changes.
    * afterChange which triggers after a slide changes.
    * afterSetup which triggers after the slider has setup completely.

### 1.4.2 (2012-07-23)
* Added fade transition.

### 1.4.1 (2012-04-21)
* Changed the arrows from &lt;span&gt;s to &lt;a&gt;s for improved accessibility.
* Added the ability to choose the starting slide.
* Some code enhancements.

### 1.4 (2012-02-02)
* Added the ability to swipe between slides.
* Increased the minimum required jQuery version to 1.5.
* Changed the following class names:
    * prev-arrow to as-prev-arrow
    * next-arrow to as-next-arrow
    * active to as-active

### 1.3.2 (2011-12-10)
* Fixed an issue where AnySlider would autoplay when there was only one slide.
* Added navigation bullets.
* Added the "rtl" option to control slide direction when autoplay is enabled.

### 1.3.1 (2011-11-01)
* Added support for custom easings via the jQuery Easing plugin.
* Added autoplay support. Check out the "interval" and "pauseOnHover" options.
* Restructured the CSS to help with the setup of AnySlider.

### 1.3 (2011-10-12)
* Rewrote most of the script.
* The slider is now "infinite" and will always scroll in the same direction.

### 1.2.2 (2011-08-29)
* Fixed an issue where the "speed" option wasn't honored.

### 1.2.1 (2011-08-28)
* AnySlider now support several sliders on one page.
* Removed the IDs from the arrow elements. Instead, they now have classes of "prev-arrow" and "next-arrow" respectivly.
* The ID "slide-inner" is now a class instead.
* Fixed an issue where animations could be "queued" and would be run several times.

### 1.2 (2011-05-10)
* Finally added a "speed" option.
* Various code improvements.

### 1.1.2 (2011-04-24)
* Enhanced the options to always show/hide arrows or just on hover.
* Minor code cleanup.

### 1.1 (2011-02-26)
* New arrows.
* Added some options:
    * Always show/hide arrows
    * Enable/disable keyboard navigation
    * Custom labels for the controls
* Fixed a minor bug were keyboard codes had to be converted into a string.
* Separated options from internal variables.

### 1.0.1 (2011-01-26)
* Rewrote bits of the code to perform better.

### 1.0 (2011-01-16)
* Initial release.


Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        copy: {
            main: {
                src: 'src/jquery.anyslider.js',
                dest: 'dist/jquery.anyslider.js'
            }
        },

        jscs: {
            src: 'src/jquery.anyslider.js'
        },

        jshint: {
            dist: {
                src: ['src/jquery.anyslider.js'],
                options: grunt.file.readJSON('.jshintrc')
            }
        },

        uglify: {
            options: {
                banner: '/*! jQuery AnySlider <%= pkg.version %> | Copyright <%= grunt.template.today("yyyy") %> Jonathan Wilsson and contributors. */\n'
            },
            build: {
                src: 'src/jquery.anyslider.js',
                dest: 'dist/jquery.anyslider.min.js'
            }
        }
    });

    grunt.loadNpmTasks('grunt-jscs');
    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-uglify');

    grunt.registerTask('default', ['jshint', 'copy', 'uglify']);
};


demo/js/jquery.easing.1.3.js
/*
 * jQuery Easing v1.3 - http://gsgd.co.uk/sandbox/jquery/easing/
 *
 * Uses the built in easing capabilities added In jQuery 1.1
 * to offer multiple easing options
 *
 * TERMS OF USE - jQuery Easing
 *
 * Open source under the BSD License.
 *
 * Copyright Â© 2008 George McGinley Smith
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list of
 * conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list
 * of conditions and the following disclaimer in the documentation and/or other materials
 * provided with the distribution.
 *
 * Neither the name of the author nor the names of contributors may be used to endorse
 * or promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 *  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
*/
jQuery.easing.jswing=jQuery.easing.swing; jQuery.extend(jQuery.easing,{def:"easeOutQuad",swing:function(e,a,c,b,d){return jQuery.easing[jQuery.easing.def](e,a,c,b,d)},easeInQuad:function(e,a,c,b,d){return b*(a/=d)*a+c},easeOutQuad:function(e,a,c,b,d){return-b*(a/=d)*(a-2)+c},easeInOutQuad:function(e,a,c,b,d){return(a/=d/2)<1?b/2*a*a+c:-b/2*(--a*(a-2)-1)+c},easeInCubic:function(e,a,c,b,d){return b*(a/=d)*a*a+c},easeOutCubic:function(e,a,c,b,d){return b*((a=a/d-1)*a*a+1)+c},easeInOutCubic:function(e,a,c,b,d){return(a/=d/2)<1?b/2*a*a*a+c:b/ 2*((a-=2)*a*a+2)+c},easeInQuart:function(e,a,c,b,d){return b*(a/=d)*a*a*a+c},easeOutQuart:function(e,a,c,b,d){return-b*((a=a/d-1)*a*a*a-1)+c},easeInOutQuart:function(e,a,c,b,d){return(a/=d/2)<1?b/2*a*a*a*a+c:-b/2*((a-=2)*a*a*a-2)+c},easeInQuint:function(e,a,c,b,d){return b*(a/=d)*a*a*a*a+c},easeOutQuint:function(e,a,c,b,d){return b*((a=a/d-1)*a*a*a*a+1)+c},easeInOutQuint:function(e,a,c,b,d){return(a/=d/2)<1?b/2*a*a*a*a*a+c:b/2*((a-=2)*a*a*a*a+2)+c},easeInSine:function(e,a,c,b,d){return-b*Math.cos(a/ d*(Math.PI/2))+b+c},easeOutSine:function(e,a,c,b,d){return b*Math.sin(a/d*(Math.PI/2))+c},easeInOutSine:function(e,a,c,b,d){return-b/2*(Math.cos(Math.PI*a/d)-1)+c},easeInExpo:function(e,a,c,b,d){return a==0?c:b*Math.pow(2,10*(a/d-1))+c},easeOutExpo:function(e,a,c,b,d){return a==d?c+b:b*(-Math.pow(2,-10*a/d)+1)+c},easeInOutExpo:function(e,a,c,b,d){return a==0?c:a==d?c+b:(a/=d/2)<1?b/2*Math.pow(2,10*(a-1))+c:b/2*(-Math.pow(2,-10*--a)+2)+c},easeInCirc:function(e,a,c,b,d){return-b*(Math.sqrt(1-(a/=d)* a)-1)+c},easeOutCirc:function(e,a,c,b,d){return b*Math.sqrt(1-(a=a/d-1)*a)+c},easeInOutCirc:function(e,a,c,b,d){return(a/=d/2)<1?-b/2*(Math.sqrt(1-a*a)-1)+c:b/2*(Math.sqrt(1-(a-=2)*a)+1)+c},easeInElastic:function(e,a,c,b,d){var e=1.70158,f=0,g=b;if(a==0)return c;if((a/=d)==1)return c+b;f||(f=d*0.3);g<Math.abs(b)?(g=b,e=f/4):e=f/(2*Math.PI)*Math.asin(b/g);return-(g*Math.pow(2,10*(a-=1))*Math.sin((a*d-e)*2*Math.PI/f))+c},easeOutElastic:function(e,a,c,b,d){var e=1.70158,f=0,g=b;if(a==0)return c;if((a/= d)==1)return c+b;f||(f=d*0.3);g<Math.abs(b)?(g=b,e=f/4):e=f/(2*Math.PI)*Math.asin(b/g);return g*Math.pow(2,-10*a)*Math.sin((a*d-e)*2*Math.PI/f)+b+c},easeInOutElastic:function(e,a,c,b,d){var e=1.70158,f=0,g=b;if(a==0)return c;if((a/=d/2)==2)return c+b;f||(f=d*0.3*1.5);g<Math.abs(b)?(g=b,e=f/4):e=f/(2*Math.PI)*Math.asin(b/g);return a<1?-0.5*g*Math.pow(2,10*(a-=1))*Math.sin((a*d-e)*2*Math.PI/f)+c:g*Math.pow(2,-10*(a-=1))*Math.sin((a*d-e)*2*Math.PI/f)*0.5+b+c},easeInBack:function(e,a,c,b,d,f){f==void 0&& (f=1.70158);return b*(a/=d)*a*((f+1)*a-f)+c},easeOutBack:function(e,a,c,b,d,f){f==void 0&&(f=1.70158);return b*((a=a/d-1)*a*((f+1)*a+f)+1)+c},easeInOutBack:function(e,a,c,b,d,f){f==void 0&&(f=1.70158);return(a/=d/2)<1?b/2*a*a*(((f*=1.525)+1)*a-f)+c:b/2*((a-=2)*a*(((f*=1.525)+1)*a+f)+2)+c},easeInBounce:function(e,a,c,b,d){return b-jQuery.easing.easeOutBounce(e,d-a,0,b,d)+c},easeOutBounce:function(e,a,c,b,d){return(a/=d)<1/2.75?b*7.5625*a*a+c:a<2/2.75?b*(7.5625*(a-=1.5/2.75)*a+0.75)+c:a<2.5/2.75?b* (7.5625*(a-=2.25/2.75)*a+0.9375)+c:b*(7.5625*(a-=2.625/2.75)*a+0.984375)+c},easeInOutBounce:function(e,a,c,b,d){return a<d/2?jQuery.easing.easeInBounce(e,a*2,0,b,d)*0.5+c:jQuery.easing.easeOutBounce(e,a*2-d,0,b,d)*0.5+b*0.5+c}});


src/jquery.anyslider.js
/*! jQuery AnySlider 2.1.0-beta | Copyright 2014 Jonathan Wilsson and contributors */

;(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD
        define(['jquery'], factory);
    } else if (typeof exports === 'object') {
        // CommonJS
        factory(require('jquery'), window, document);
    } else {
        factory(root.jQuery || root.Zepto, window, document);
    }
}(this, function ($, window, document) {
    'use strict';

    var AnySlider = function (slider, options) {
        var slides = slider.children();
        var orgNumSlides = slides.length;
        var numSlides = orgNumSlides;
        var width = slider.width();
        var nextSlide = 0;
        var current = 0;
        var inner;
        var timer;
        var running = false;
        var defaults = {
                afterChange: function () {},
                afterSetup: function () {},
                animation: 'slide',
                beforeChange: function () {},
                easing: 'swing',
                interval: 5000,
                keyboard: true,
                nextLabel: 'Next slide',
                pauseOnHover: true,
                prevLabel: 'Previous slide',
                reverse: false,
                showBullets: true,
                showControls: true,
                speed: 400,
                startSlide: 1,
                touch: true
            };

        options = $.extend(defaults, options);

        // Setup the slides
        if (orgNumSlides > 1) {
            slides.eq(0).clone().addClass('clone').appendTo(slider);
            slides.eq(numSlides - 1).clone().addClass('clone').prependTo(slider);

            if (options.startSlide < orgNumSlides) {
                current = options.startSlide;
            }
        }

        slides = slider.children();
        numSlides = slides.length;

        // CSS setup
        slides.wrapAll('<div class="as-slide-inner"></div>').css('width', width);
        inner = slider.css('overflow', 'hidden').find('.as-slide-inner');

        if (options.animation === 'fade') {
            // Properties are quoted for consistency since "float" will trigger an error when the script is minified (if unquoted)
            slides.css({
                'display': 'none',
                'left': 0,
                'position': 'absolute',
                'top': 0
            }).eq(current).show();

            inner.css('width', width);
        } else {
            slides.css({
                'float': 'left',
                'position': 'relative'
            });

            inner.css({
                'left': -current * width,
                'width': numSlides * width
            });
        }

        inner.css({
            'float': 'left',
            'position': 'relative'
        });

        // Add the arrows
        if (options.showControls && orgNumSlides > 1) {
            slider.prepend('<a href="#" class="as-prev-arrow" title="LABEL">LABEL</a>'.replace(/LABEL/g, options.prevLabel));
            slider.append('<a href="#" class="as-next-arrow" title="LABEL">LABEL</a>'.replace(/LABEL/g, options.nextLabel));

            slider.on('click.as', '.as-prev-arrow, .as-next-arrow', function (e) {
                e.preventDefault();

                if (running) {
                    return;
                }

                if ($(this).hasClass('as-prev-arrow')) {
                    prev();
                } else {
                    next();
                }
            });
        }

        // Add navigation bullets
        if (options.showBullets && orgNumSlides > 1) {
            var out = '<div class="as-nav"></div>';
            var nav = $(out);
            var active;
            var i;

            for (i = 1; i <= orgNumSlides; i++) {
                active = '';

                if (i === current) {
                    active = ' class="as-active"';
                }

                nav.append('<a href="#"' + active + '>' + i + '</a>');
            }

            nav.on('click.as', 'a', function (e) {
                var index = $(this).index();

                e.preventDefault();

                if ($(this).hasClass('as-active') || running) {
                    return;
                }

                nav.find('a').removeClass('as-active').eq(index).addClass('as-active');

                goTo(index + 1);
            });

            slider.after(nav);
        }

        // Enable keyboard navigation
        if (options.keyboard) {
            $(document).on('keydown.as', function (e) {
                var key = e.keyCode;

                // See if the left or right arrow is pressed
                if (key !== 37 && key !== 39 || orgNumSlides <= 1) {
                    return;
                }

                if (key === 37) {
                    prev();
                } else {
                    next();
                }
            });
        }

        if (options.pauseOnHover) {
            slider.on('mouseenter', function () {
                pause();
            }).on('mouseleave', function () {
                play();
            });
        }

        // Enable responsive support
        $(window).resize(function () {
            if (!running) {
                width = slider.width();

                inner.css('width', width);
                slides.css('width', width);

                if (options.animation !== 'fade') {
                    inner.css({
                        'left': -current * width,
                        'width': numSlides * width
                    });
                }
            }
        });

        /**
         * Enable swipe support. (Don't bother if not wanted or supported)
         *
         * Resources:
         * http://wowmotty.blogspot.com/2011/10/adding-swipe-support.html
         * https://blogs.windows.com/buildingapps/2012/11/15/adapting-your-webkit-optimized-site-for-internet-explorer-10/#step4
         * https://hacks.mozilla.org/2013/04/detecting-touch-its-the-why-not-the-how/
         */
        if (options.touch && ('ontouchstart' in window) || (navigator.msMaxTouchPoints > 0) || (navigator.maxTouchPoints > 0)) {
            var startTime;
            var startX;

            slider.on('touchstart.as pointerdown.as MSPointerDown.as', function (e) {
                startTime = e.timeStamp;

                if (e.originalEvent) {
                    startX = e.originalEvent.pageX || e.originalEvent.touches[0].pageX;
                } else {
                    startX = e.pageX || e.touches[0].pageX;
                }
            }).on('touchmove.as pointermove.as MSPointerMove.as', function (e) {
                var currentTime = e.timeStamp;
                var currentDistance = 0;
                var currentX;

                if (e.originalEvent) {
                    currentX = e.originalEvent.pageX || e.originalEvent.touches[0].pageX;
                } else {
                    currentX = e.pageX || e.touches[0].pageX;
                }

                if (startX !== 0) {
                    currentDistance = Math.abs(currentX - startX);
                }

                // Only allow if movement < 1 sec and distance is long enough
                if (startTime !== 0 && currentTime - startTime < 1000 && currentDistance > 10) {
                    e.preventDefault();

                    if (currentX < startX) { // Swiping to the left, i.e. next slide
                        next();
                    } else if (currentX > startX) { // Swiping to the right, i.e. previous slide
                        prev();
                    }

                    startTime = 0;
                    startX = 0;

                    // Android doesn't always fire touchend
                    slider.trigger('touchend.as');
                }
            }).on('touchend.as pointerup.as MSPointerUp.as', function () {
                startTime = 0;
                startX = 0;
            });
        }

        // Enable autoplay
        tick();

        options.afterSetup.call(slider[0]);

        // Private methods
        // Animation complete callback
        function animationCallback () {
            current = nextSlide;

            if (nextSlide === 0) {
                current = orgNumSlides;

                if (options.animation !== 'fade') {
                    inner.css('left', -current * width);
                }
            } else if (nextSlide === numSlides - 1) {
                current = 1;

                if (options.animation !== 'fade') {
                    inner.css('left', -width);
                }
            }

            // Fix for Zepto hiding the slide
            if (options.animation === 'fade') {
                slides.eq(current).show();
            }

            if (options.showBullets) {
                slider.next('.as-nav').find('a').removeClass('as-active').eq(current - 1).addClass('as-active');
            }

            running = false;

            options.afterChange.call(slider[0]);
        }

        // The main animation function
        function run () {
            if (running || orgNumSlides <= 1) {
                return;
            }

            running = true;

            options.beforeChange.call(slider[0]);

            if (options.animation === 'fade') {
                slides.css('z-index', 1).fadeOut(options.speed).eq(nextSlide).css('z-index', 2).fadeIn(options.speed, animationCallback);
            } else {
                inner.animate({'left': -nextSlide * width}, options.speed, options.easing, animationCallback);
            }

            tick();
        }

        // Set the autoplay timer
        function tick () {
            clearTimeout(timer);

            // Check if autoplay is enabled
            if (options.interval && orgNumSlides > 1) {
                timer = setTimeout(function () {
                    if (options.reverse) {
                        prev();
                    } else {
                        next();
                    }
                }, options.interval);
            }
        }

        // Public methods
        function currentSlide () {
            return current;
        }

        function goTo (slide) {
            nextSlide = slide;

            run();
        }

        function next () {
            nextSlide = current + 1;

            run();
        }

        function pause () {
            clearTimeout(timer);
        }

        function play () {
            tick();
        }

        function prev () {
            nextSlide = current - 1;

            run();
        }

        return {
            currentSlide: currentSlide,
            goTo: goTo,
            next: next,
            pause: pause,
            play: play,
            prev: prev
        };
    };

    $.fn.anyslider = function (options) {
        return this.each(function () {
            var slider = $(this);
            var anyslider;

            // Bail if we already have a plugin instance for this element
            if (slider.data('anyslider')) {
                return slider.data('anyslider');
            }

            anyslider = new AnySlider(slider, options);

            slider.data('anyslider', anyslider);
        });
    };
}));
